File: front/src/app/items-factura/itemFactura.ts
import {Producto} from '../productos/producto';
import {OnInit} from '@angular/core';

export class ItemFactura implements OnInit {
  private _id: number;
  private _producto: Producto;
  private _precio: number;
  private _cantidad: number;
  private _importe: number;

  constructor(producto: Producto) {
    this._producto = producto;
    this._cantidad = 1;
    this._precio = producto.precio;
    this.setImporte();
  }

  ngOnInit(): void {
  }

  get id(): number {
    return this._id;
  }

  set id(value: number) {
    this._id = value;
  }

  get producto(): Producto {
    return this._producto;
  }

  set producto(value: Producto) {
    this._producto = value;
  }

  get cantidad(): number {
    return this._cantidad;
  }

  set cantidad(value: number) {
    this._cantidad = value;
  }

  get importe(): number {
    return this._importe;
  }

  set importe(value: number) {
    this._importe = value;
  }

  setImporte() {
    return this._importe = this._precio * this._cantidad;
  }

  get precio(): number {
    return this._precio;
  }

  set precio(value: number) {
    this._precio = value;
  }

  toJSON() {
    return {
      id: this.id,
      producto: this.producto,
      precio: this.precio,
      cantidad: this.cantidad,
      importe: this.importe
    };
  }
}

File: back/src/main/java/com/curso/spring/rest/models/entity/ItemFactura.java
package com.curso.spring.rest.models.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import java.io.Serializable;

@Entity
@Table(schema = "tfg", name = "facturas_items")
public class  ItemFactura implements Serializable {

    private static final long serialVersionUID = -1142856044917006208L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JsonIgnoreProperties(value = {"hibernateLazyInitializer", "handler"})
    private Producto producto;

    private Double precio;

    private Integer cantidad;

    private Double importe;

    public ItemFactura() {}

    public ItemFactura(Integer cantidad, Producto producto) {
        this.cantidad = cantidad;
        this.producto = producto;
        this.precio = producto.getPrecio();
        importe = cantidad * precio;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Integer getCantidad() {
        return cantidad;
    }

    public void setCantidad(Integer cantidad) {
        this.cantidad = cantidad;
    }

    public Double getImporte() {
        if (importe != null)
            return this.importe;
        return cantidad * producto.getPrecio();
    }

    public void setImporte(Double importe) {
        this.importe = importe;
    }

    public Producto getProducto() {
        return producto;
    }

    public void setProducto(Producto producto) {
        this.producto = producto;
    }

    public Double getPrecio() {
        return precio;
    }

    public void setPrecio(Double precio) {
        this.precio = precio;
    }
}
